### DISCUSSION TITLE
How to write own monad without runtime info in context ?

### DISCUSSION THREAD

**Discussion by mio-19** [rssh/dotty-cps-async#91] [Q&A] [answered]
I have a stateful non-deterministic monad originally defined like this
```scala
trait Trying[State, +Result] {
  def apply(state: State): Vector[Res[State, Result]]
}
final class CpsMonadTrying[State] extends CpsMonad[[X] =>> Trying[State, X]] with CpsMonadContext[[X] =>> Trying[State, X]]
```
I can use it with async await fine.
I try to change it to
```scala
type Trying[State, +Result] = State => Vector[Res[State, Result]]
```

Then I got complains `Async function with arity != 1 is not supported yet`

When I am writing this CpsMonad I don't know what is CpsMonadContext for I just make up something to make it compile. I can't find any code or documents about CpsMonadContext  to make myself understand
```scala
final class CpsMonadTrying[State] extends CpsMonad[[X] =>> Trying[State, X]] with CpsMonadContext[[X] =>> Trying[State, X]] {
...
  override def monad: CpsMonad[WF] = this

  override type Context = this.type

}
```

**Comment by rssh** [comment, marked as answer]
About CpsMonadContext - look at this section:  https://rssh.github.io/dotty-cps-async/MonadContexts.html 
If you have no runtime infrastructure around context,  then
```
final class CpsMonadTrying[State] extends CpsMonad[[X] =>> Trying[State, X]] 
                                                                 with CpsMonadInstanceContext[[X] =>> Trying[State, X]] {
...
   // note - CpsMonadInstanceContext is a type which should be mixed,  not CpsMonadContext
}
```

Full example can be found ... 
https://github.com/rssh/dotty-cps-async/blob/master/shared/src/test/scala/cpstest/CpsContT.scala
(but it's too long,  searching shorter ...

**Reply by rssh** [reply]
Also long, but better fit to non-detemenism theme:
https://github.com/rssh/dotty-cps-async/blob/626ffb64939635b60846df13fb514027fd52213d/shared/src/test/scala/logic/logict/LogicStreamT.scala#L256

**Comment by rssh** [comment]
Also look at the logic module, https://github.com/rssh/dotty-cps-async/tree/master/logic
  where we have a few such monads defined:

https://github.com/rssh/dotty-cps-async/tree/master/logic/shared/src/main/scala/cps/monads/logic
